controllers/common/types_test.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/config"
+	"github.com/chaos-mesh/chaos-mesh/pkg/webhook/config/watcher"
+)
+
+func TestTargetNamespaceShouldSetWhileClusterScopedIsFalse(t *testing.T) {
+	c := config.ChaosControllerConfig{
+		WatcherConfig: &watcher.Config{
+			ClusterScoped:     false,
+			TemplateNamespace: "",
+			TargetNamespace:   "",
+			TemplateLabels:    nil,
+			ConfigLabels:      nil,
+		},
+		ClusterScoped:   false,
+		TargetNamespace: "",
+	}
+	err := validate(&c)
+	assert.NotNil(t, err)
+}
+
+func TestWatcherConfigIsAlwaysRequired(t *testing.T) {
+	c := config.ChaosControllerConfig{
+		WatcherConfig:   nil,
+		ClusterScoped:   true,
+		TargetNamespace: "",
+	}
+	err := validate(&c)
+	assert.NotNil(t, err)
+}
+
+func TestClusterScopedInConfigAndClusterScopedInWatcherConfigShouldKeepConstant(t *testing.T) {
+	c := config.ChaosControllerConfig{
+		WatcherConfig: &watcher.Config{
+			ClusterScoped:     false,
+			TemplateNamespace: "",
+			TargetNamespace:   "",
+			TemplateLabels:    nil,
+			ConfigLabels:      nil,
+		},
+		ClusterScoped:   true,
+		TargetNamespace: "",
+	}
+	err := validate(&c)
+	assert.NotNil(t, err)
+}
+
+func TestNsInConfigAndNsInWatcherConfigShouldKeepConstant(t *testing.T) {
+	c := config.ChaosControllerConfig{
+		WatcherConfig: &watcher.Config{
+			ClusterScoped:   false,
+			TargetNamespace: "ns1",
+		},
+		ClusterScoped:   false,
+		TargetNamespace: "ns2",
+	}
+	err := validate(&c)
+	assert.NotNil(t, err)
+}
```

Comment:
I think these 4 test cases could be merged into one test suite and managed by table-driven style.

Reply:
OK. I will migrate these testcases with ginkgo. :)

Reply:
Well,I think what @Yisaer  meant was look  like this https://github.com/chaos-mesh/chaos-mesh/blob/5d409d77ae86c3896b15e2809c6b977999368479/api/v1alpha1/iochaos_webhook_test.go#L35-L40

Reply:
Oops, I have a misunderstanding on table-driven style. 😅

I will refactor these test cases right now.

Reply:
@Gallardot PTAL. :)